### PR TITLE
feat(gamma): Shared Policy Group lifecycle and history

### PR DESCRIPTION
**Pull request by ankita-gupta21**
## Issue
https://gravitee.atlassian.net/browse/FOUND-4

## Description
Brings Shared Policy Group management in Gamma to Classic Console parity.

- Adds lifecycle controls for save, deploy, undeploy, and delete, with save-before-deploy behavior, permission-aware actions, and Kubernetes-origin read-only handling.
- Replaces detail tabs with a single actions menu while retaining primary Save and Deploy controls.
- Adds policy-step enable/disable controls, visible conditions, and wrapping for long policy chains.
- Adds version history with JSON source, a read-only historical policy studio, version comparison, comparison with pending changes, and restore.
- Reuses a shared delete confirmation across list and detail views.

attached Working screenshot: https://gravitee.atlassian.net/browse/FOUND-4?focusedCommentId=102646

### CODE REVIEW

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupPolicyStudio.tsx [hunk 1]
```diff
@@ -136,50 +154,75 @@ export function SharedPolicyGroupPolicyStudio({
...
+                            variant="outline"
+                            disabled={!hasChanges || isSaving}
+                            onClick={() => void handleSave()}
+                        >
                             {isSaving ? 'Saving…' : 'Save policies'}
                         </Button>
+                        <Button
+                            type="button"
+                            size="sm"
+                            disabled={sharedPolicyGroup.lifecycleState === 'DEPLOYED' || hasChanges || lifecyclePending}
```
Comment (aqibmohammadkhan) [inline gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupPolicyStudio.tsx]: Deploy button stuck disabled after Save.
a stale-query race where the UI still reads lifecycleState: 'DEPLOYED' right after saving, even though the backend already flipped it to PENDING.

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/services/sharedPolicyGroups.ts [hunk 2]
```diff
@@ -66,3 +66,39 @@ export async function deleteSharedPolicyGroup(environmentId: string, sharedPolic
...
+        searchParams.set('sortBy', params.sortBy);
+    }
+    return apimFetchJsonV2<SharedPolicyGroupsPagedResponse>(
+        environmentId,
+        `/shared-policy-groups/${encodeURIComponent(sharedPolicyGroupId)}/histories?${searchParams.toString()}`,
+    );
+}
+
+export async function restoreSharedPolicyGroup(environmentId: string, history: SharedPolicyGroup): Promise<SharedPolicyGroup> {
+    return apimFetchJsonV2<SharedPolicyGroup>(environmentId, `/shared-policy-groups/${encodeURIComponent(history.id)}`, {
```
Comment (aqibmohammadkhan) [inline gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/services/sharedPolicyGroups.ts:107]: Restore sends full historical object (incl. lifecycleState) to update endpoint.
plausibly leaves the group stuck showing DEPLOYED right when the UI tells the user to click Deploy to finish restoring.

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.tsx [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                        <DropdownMenuItem disabled={undeployMutation.isPending} onSelect={() => void handleUndeploy()}>
+                            <RocketIcon className="mr-2 size-4" aria-hidden />
+                            {undeployMutation.isPending ? 'Undeploying…' : 'Undeploy'}
+                        </DropdownMenuItem>
+                    ) : null}
+                    <DropdownMenuItem onSelect={() => navigate(historyHref)}>
+                        <ClockIcon className="mr-2 size-4" aria-hidden />
+                        Version History
+                    </DropdownMenuItem>
+                    {canUpdate && canDelete && canMutate ? (
```
Comment (aqibmohammadkhan) [inline gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.tsx:122]: Delete permission is inconsistent between the list page and the detail page's actions menu. a user with delete-only permission can delete from one surface but not the other.

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupDetailLayout.tsx
Comment (aqibmohammadkhan) [inline gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupDetailLayout.tsx:1]: Prerequisite message / last-updated / last-deployed no longer shown.

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryCompareDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    const leftJson = JSON.stringify(left, null, 2);
+    const rightJson = JSON.stringify(right, null, 2);
+    const changes = diffLines(leftJson, rightJson);
+    const rightVersionLabel = rightIsPending ? 'to be deployed' : (right.version ?? '—');
+    const title = `Comparing version ${left.version ?? '—'} with version ${rightVersionLabel}`;
+
+    return (
+        <Dialog open={open} onOpenChange={onOpenChange}>
+            <DialogContent className="w-[min(72rem,calc(100vw-2rem))] max-w-none">
```
Comment (aqibmohammadkhan) [inline gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryCompareDialog.tsx]: Two history dialogs (Compare, JSON Source) collapse to a narrow, unreadable width — they're missing the sm:max-w-none fix that the sibling Details dialog already has.

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryDetailsDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        queryFn: listPolicies,
+        staleTime: 5 * 60 * 1000,
+    });
+    const onFetchPolicySchema = useCallback((policy: Policy) => getPolicySchema(policy.id, protocolType), [protocolType]);
+    const onFetchPolicyDocumentation = useCallback((policy: Policy) => getPolicyDocumentation(policy.id, protocolType), [protocolType]);
+
+    return (
+        <Dialog open onOpenChange={onOpenChange}>
+            {/* DialogContent caps itself at `sm:max-w-sm`. An unprefixed `max-w-none` never wins against a
+                breakpoint-prefixed class, so the cap has to be cleared at that breakpoint too — without
```
Comment (aqibmohammadkhan) [inline gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryDetailsDialog.tsx:89]: History details dialog key can collide across rows

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.spec.tsx
Comment (aqibmohammadkhan) [inline gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.spec.tsx:1]: No test covers edit-failure error toast/sheet behavior

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryCompareDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    right,
+    rightIsPending = false,
+    onOpenChange,
+}: SharedPolicyGroupHistoryCompareDialogProps) {
+    if (!left || !right) {
+        return null;
+    }
+
+    const leftJson = JSON.stringify(left, null, 2);
+    const rightJson = JSON.stringify(right, null, 2);
```
Comment (aqibmohammadkhan) [inline gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryCompareDialog.tsx]: History compare diff recomputes on every render (no memoization)

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupPolicyStudio.tsx (hunk 1 above)
Comment (ankita-gupta21) [reply gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupPolicyStudio.tsx]: After save, the studio still used a cached DEPLOYED lifecycle, so Deploy stayed disabled. It now uses the update response (PENDING) so Deploy enables right away.

[State handling L87–L91](https://github.com/gravitee-io/gravitee-api-management/blob/827948bbbb59e8cefd0c3dc02ea1459661c9e71d/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupPolicyStudio.tsx#L87-L91)
[Save response L134–L144](https://github.com/gravitee-io/gravitee-api-management/blob/827948bbbb59e8cefd0c3dc02ea1459661c9e71d/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupPolicyStudio.tsx#L134-L144)

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/services/sharedPolicyGroups.ts (hunk 2 above)
Comment (ankita-gupta21) [reply gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/services/sharedPolicyGroups.ts:107]: Restore posted the historical object, including lifecycleState, so the group could stay DEPLOYED while the UI asked you to Deploy. Restore now sends only mutable update fields, not lifecycleState.

[Restore payload L99–L110](https://github.com/gravitee-io/gravitee-api-management/blob/827948bbbb59e8cefd0c3dc02ea1459661c9e71d/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/services/sharedPolicyGroups.ts#L99-L110)

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.tsx (hunk 3 above)
Comment (ankita-gupta21) [reply gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.tsx:122]: A delete-only user could delete from the list but not from the detail menu. Both surfaces now require update and delete.

[Detail L122–L130](https://github.com/gravitee-io/gravitee-api-management/blob/827948bbbb59e8cefd0c3dc02ea1459661c9e71d/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.tsx#L122-L130)
[List L122–L126](https://github.com/gravitee-io/gravitee-api-management/blob/827948bbbb59e8cefd0c3dc02ea1459661c9e71d/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupsTable.tsx#L122-L126)

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupDetailLayout.tsx
Comment (ankita-gupta21) [reply gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupDetailLayout.tsx:1]: Detail header again shows prerequisite message and timestamps via DateCell.

[L44–L71](https://github.com/gravitee-io/gravitee-api-management/blob/c5ab720c0aa10c50ed04082b0ee63d7ec173c69a/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupDetailLayout.tsx#L44-L71)

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryCompareDialog.tsx (hunk 4 above)
Comment (ankita-gupta21) [reply gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryCompareDialog.tsx]: Both use sm:max-w-none, same as Details.

[Compare L87–L88](https://github.com/gravitee-io/gravitee-api-management/blob/c5ab720c0aa10c50ed04082b0ee63d7ec173c69a/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryCompareDialog.tsx#L87-L88)
[JSON Source L40–L42](https://github.com/gravitee-io/gravitee-api-management/blob/c5ab720c0aa10c50ed04082b0ee63d7ec173c69a/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryJsonDialog.tsx#L40-L42)

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryDetailsDialog.tsx (hunk 5 above)
Comment (ankita-gupta21) [reply gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryDetailsDialog.tsx:89]: Key is now id + version.

[L57–L60](https://github.com/gravitee-io/gravitee-api-management/blob/c5ab720c0aa10c50ed04082b0ee63d7ec173c69a/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryDetailsDialog.tsx#L57-L60)

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.spec.tsx
Comment (ankita-gupta21) [reply gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.spec.tsx:1]: Spec asserts error toast and that the edit sheet stays open.

[L128–L139](https://github.com/gravitee-io/gravitee-api-management/blob/c5ab720c0aa10c50ed04082b0ee63d7ec173c69a/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.spec.tsx#L128-L139)

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryCompareDialog.tsx (hunk 6 above)
Comment (ankita-gupta21) [reply gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryCompareDialog.tsx]: JSON stringify and diffLines are memoized.

[L72–L77](https://github.com/gravitee-io/gravitee-api-management/blob/c5ab720c0aa10c50ed04082b0ee63d7ec173c69a/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryCompareDialog.tsx#L72-L77)

**Review by aqibmohammadkhan — COMMENTED** [review]
1. Deploy button gets stuck disabled right after Save (SharedPolicyGroupPolicyStudio.tsx:170) — this undermines the PR's headline feature, "save-before-deploy behavior." A user saves, then can't deploy until an unrelated re-render happens to trigger a refetch.
2. Delete permission is inconsistent between list and detail views (SharedPolicyGroupActions.tsx:122) — the PR advertises "permission-aware actions," but the same user gets a different authorization outcome depending on which screen they're on. That's a real access-control bug, not just a UX inconsistency.
3. Restore likely can't be deployed afterward (PLAUSIBLE — sends lifecycleState back on restore, which may re-trigger the same stuck-Deploy-button bug) — worth confirming against the backend behavior before merging, since it's core to the new "restore" feature this PR adds.
4. Prerequisite message / last-updated / last-deployed silently no longer visible anywhere on the detail page — a real regression in visible information, even if not a crash.
5. History Compare/JSON dialogs render at ~24rem wide (unreadable) — visible on first use, easy to demo-fail.
6. Dropped edit-failure test coverage — not a bug itself, but removes a guard against future regressions in error handling.
7. Duplicated lineDiff.ts, unmemoized diff recompute, possible history-dialog key collision on edge-case data.

**Comment by ankita-gupta21** [comment]
> 1. Deploy button gets stuck disabled right after Save (SharedPolicyGroupPolicyStudio.tsx:170) — this undermines the PR's headline feature, "save-before-deploy behavior." A user saves, then can't deploy until an unrelated re-render happens to trigger a refetch.
> 2. Delete permission is inconsistent between list and detail views (SharedPolicyGroupActions.tsx:122) — the PR advertises "permission-aware actions," but the same user gets a different authorization outcome depending on which screen they're on. That's a real access-control bug, not just a UX inconsistency.
> 3. Restore likely can't be deployed afterward (PLAUSIBLE — sends lifecycleState back on restore, which may re-trigger the same stuck-Deploy-button bug) — worth confirming against the backend behavior before merging, since it's core to the new "restore" feature this PR adds.
> 4. Prerequisite message / last-updated / last-deployed silently no longer visible anywhere on the detail page — a real regression in visible information, even if not a crash.
> 5. History Compare/JSON dialogs render at ~24rem wide (unreadable) — visible on first use, easy to demo-fail.
> 6. Dropped edit-failure test coverage — not a bug itself, but removes a guard against future regressions in error handling.
> 7. Duplicated lineDiff.ts, unmemoized diff recompute, possible history-dialog key collision on edge-case data.

1. Deploy stuck after Save — UI still thought the group was DEPLOYED. Save now uses the update response (PENDING). [Studio L134–L145](https://github.com/gravitee-io/gravitee-api-management/blob/e31279b064/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupPolicyStudio.tsx#L134-L145)
Before saving — Save is enabled, Deploy is disabled because the draft has unsaved edits:
image
After one click on Save — the badge flips to Pending, "Last updated" advances to 19:41, Save greys out and Deploy is immediately enabled. No second save, no refetch needed:
image
After clicking Deploy — badge is Deployed, both buttons disabled, "Last deployed" matches "Last updated" at 19:42: 
image

2. Delete list vs detail — delete-only could delete from the list. Both surfaces now need update and delete. [Detail L122](https://github.com/gravitee-io/gravitee-api-management/blob/e31279b064/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.tsx#L122) · [List L125](https://github.com/gravitee-io/gravitee-api-management/blob/e31279b064/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupsTable.tsx#L125) 
image
image

3. Restore sent lifecycleState — that could keep Deploy disabled. Restore sends only update fields. [L99–L110](https://github.com/gravitee-io/gravitee-api-management/blob/e31279b064/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/services/sharedPolicyGroups.ts#L99-L110) 
image

4. Missing prerequisite / timestamps on detail — restored in the header. [L44–L71](https://github.com/gravitee-io/gravitee-api-management/blob/e31279b064/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupDetailLayout.tsx#L44-L71) 

5. Compare/JSON ~24rem wide — sm:max-w-none on both dialogs. [Compare L88](https://github.com/gravitee-io/gravitee-api-management/blob/e31279b064/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryCompareDialog.tsx#L88) · [JSON L42](https://github.com/gravitee-io/gravitee-api-management/blob/e31279b064/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupHistoryJsonDialog.tsx#L42) 
image
image


6. No edit-failure test — [Actions spec L128–L139](https://github.com/gravitee-io/gravitee-api-management/blob/e31279b064/gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/shared-policy-groups/components/SharedPolicyGroupActions.spec.tsx#L128-L139)
7. Unmemoized diff / unstable dialog key — useMemo on compare; details key is id-version. Copied lineDiff.ts stays (same pattern as module-apim; not a shared package).
image
image

**Review by aqibmohammadkhan — APPROVED** [review]
